### PR TITLE
Set expand=False by default in discrete enumeration

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #915, #1337 

This PR flips our default behavior of discrete enumeration from `expand=True` to `expand=False`.

While updating the tensor shapes tutorial, I've also rewritten to rely more heavily on `poutine.broadcast` and to avoid mention of `.expand_by()` in favor of `.expand()`.

## Tasks

- [x] Simplify existing unit tests
- [x] Update tensor shapes tutorial
- [x] Fix internal applications to set `expand=True` to avoid breakage

### CODE REVIEW

File: tutorial/source/tensor_shapes.ipynb [hunk 1]
```diff
@@ -530,7 +535,7 @@
    "source": [
     "### Automatic broadcasting via broadcast poutine<a class=\"anchor\" id=\"Automatic-broadcasting-via-broadcast-poutine\"></a>\n",
     "\n",
-    "Note that in all our model/guide specifications, we had to expand sample shapes by hand to satisfy the constraints on batch shape enforced by `pyro.iarange` statements. This code can be simplified by using [poutine.broadcast](http://docs.pyro.ai/en/latest/poutine.html#pyro.poutine.broadcast), which automatically broadcasts the batch shape of `pyro.sample` statements when inside a single or nested iarange context. \n",
+    "Note that in all our model/guide specifications, we hve relied on [poutine.broadcast](http://docs.pyro.ai/en/latest/poutine.html#pyro.poutine.broadcast) to automatically expand sample shapes to satisfy the constraints on batch shape enforced by `pyro.iarange` statements. However `poutine.broadcast` is equavlent to hand-annotated `.expand()` statements.\n",
```
Comment (neerajprad) [inline tutorial/source/tensor_shapes.ipynb]: Check --> `have`, `equivalent`.  👍 description change!

File: tutorial/source/tensor_shapes.ipynb [hunk 2]
```diff
@@ -273,12 +274,12 @@
     "    x_axis = pyro.iarange(\"x_axis\", 3, dim=-2)\n",
     "    y_axis = pyro.iarange(\"y_axis\", 2, dim=-3)\n",
     "    with x_axis:\n",
-    "        x = pyro.sample(\"x\", Normal(0, 1).expand_by([3, 1]))\n",
+    "        x = pyro.sample(\"x\", Normal(0, 1))\n",
```
Comment (neerajprad) [inline tutorial/source/tensor_shapes.ipynb]: Are we relying on the automatic broadcasting in `TraceEnum_ELBO` here?

Comment (fritzo) [reply tutorial/source/tensor_shapes.ipynb]: Oops, yes, but I'll add an explicit `@poutine.broadcast`

Comment (neerajprad) [reply tutorial/source/tensor_shapes.ipynb]: We can also link to the section below for the out of order appearance of `poutine.broadcast`.

**Pull request merged**